### PR TITLE
ci: release workflow

### LINKED ISSUE #218: Make deployment consistent with other apps
Production release should trigger only upon creating a new release, which means we also need to integrate optic (release automation action) in this repo.

Inspire from how it's done e.g. in https://github.com/nearform/optic-release-automation-app

### PR DESCRIPTION
**Pull request by Pehesi97**
This PR:

- Uses the `nearform/optic-release-automation` as the release workflow for the repo
- Uses the manually created GCP OIDC configuration to authenticate when running the CD workflow
- Uses GitHub Actions Secrets as the single source of truth for secrets in the repo

In the CD workflow, the changes were to create / update the secrets in Google Secret Manager using the values stored in GitHub, and to make the Cloud Run service pull those secrets from Secret Manager on runtime so the values won't show up plain-text in the GCP Dashboard.

Closes #218.

### CODE REVIEW

**Comment by Pehesi97** [comment]
@simoneb @zimny if I understood the task well, I guess that's it, right? This doesn't look like an app we publish to npm, so no npm-token. Again, I might be wrong. Please correct me if so.

I guess I'm unable to test it until we merge this PR into master.

**Review by simoneb — COMMENTED** [review]
Thanks for this. Apologies that the original issue was missing some details. There are a few things that we need to do here, some of which you have done and some not. Please use this list as a reference for the other repo as well:

- we need to use optic to make releases by introducing the release workflow. please inspire/copy from how it's done in other repos were we're doing something similar. you may want to search across the org for workflows in which we use the deploy-cloudrun action, so you can see how both the deployment (cd.yml in this repo) and the release workflow are configured
- we need to switch to using OIDC instead of long lived credentials to authenticate to the cloud provider
- ideally, we should also switch from using env vars to provide secrets to the app to using gcloud secrets (I appreciate this is more changes than originally intended)

**Comment by Pehesi97** [comment]
@simoneb could you check my PR again?

One note, I don't have the permissions to enable the Secret Manager API on the Optic project in GCP. Could you either give me the permissions or create the `optic-firebase-private-key` secret there for me?

Also, we need to ensure that we have the `secrets.gcp_sa` and `secrets.gcp_provider_name` secrets created here in GitHub so the GCP auth action doesn't fail. I don't have the permissions to add secrets here either.

**Comment by Pehesi97** [comment]
@simoneb, new changes:

- [x] OIDC config manually created in the Optic GCP Project
- [x] Added a step that removes, creates and updates the secrets referenced in the action (I did that because if I didn't, we'd be creating a new version of those secrets at each workflow run, and they cost money - tell me if you want me to make this run differently)
- [x] Used the `secrets` config of the `deploy-cloudrun` action for referencing the newly created secrets

I believe this closes out the action items for this PR. Please let me know if I get anything wrong or if you need me to make more changes.

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+          - minor
+          - major
+  pull_request:
+    types: [closed]
+
+jobs:
+  release:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: nearform/optic-release-automation@main
```
Comment (simoneb) [inline .github/workflows/release.yml:22]: reference latest major instead of main please

Comment (Pehesi97) [reply .github/workflows/release.yml:22]: oh right

File: .github/workflows/cd.yml [hunk 2]
```diff
@@ -15,19 +16,44 @@ jobs:
...
+            gcloud secrets delete --quiet $secret || true
+
+            # re-create the secret
+            gcloud secrets create --locations=${{ secrets.GCP_CLOUDRUN_SERVICE_REGION }} --replication-policy=user-managed $secret
+          done
+
+          # update the secrets
+          echo "${{ secrets.FIREBASE_CLIENT_EMAIL }}" | tr -d '\n' | gcloud secrets versions add optic-firebase-client-email --data-file=-
+          echo "${{ secrets.FIREBASE_PROJECT_ID }}" | tr -d '\n' | gcloud secrets versions add optic-firebase-project-id --data-file=-
+          echo "${{ secrets.FIREBASE_PRIVATE_KEY_BASE64 }}" | tr -d '\n' | gcloud secrets versions add optic-firebase-private-key --data-file=-
```
Comment (simoneb) [inline .github/workflows/cd.yml:46]: looking at this more closely, this is slightly risky. say that something breaks when we delete the secrets and before we have a chance to create them. this will cause the existing revision to fail, wouldn't it? can we see if we can come up with a more robust way to handle this? (note: I do appreciate this way of updating secrets values via CLI is not ideal)

File: infrastructure/variables.tf [hunk 3]
```diff
@@ -13,3 +13,11 @@ variable "region" {
...
   default = "europe-west1-b"
 }
+
+variable "secrets" {
+  default = {
+    "optic-firebase-project-id"   = "FIREBASE_PROJECT_ID"
+    "optic-firebase-client-email" = "FIREBASE_CLIENT_EMAIL"
+    "optic-firebase-private-key"  = "FIREBASE_PRIVATE_KEY_BASE64"
+  }
+}
```
Comment (simoneb) [inline infrastructure/variables.tf:23]: Q: if we are using tf to provision the secrets, do we really need to create them in the cd workflow? if we don't, then the code in the workflow is largely simplified as we only have to update the value, instead of deleting and creating each secret

Comment (Pehesi97) [reply infrastructure/variables.tf:23]: I couldn't find a workflow that was supposed to run terraform. That's why I didn't create the secrets in the terraform code, but rather assumed they would be created by the time somebody runs terraform (if we tried to create a secret in terraform that is already there, terraform would fail - so people would need to import the secrets manually when running terraform). But sure we can do the other way around.

My point (and this is in answer to both comments) is that we didn't want to keep adding new secret versions since they cost money and they can't be updated. What Piotr suggested is that we could retrieve the secrets on the workflow, check their value, and if there are no changes we would just let the workflow continue. If there are changes, we create a new version of the secret. After that, we could destroy the old one.

So, to summarize:

- Create the secrets in terraform
- Only create new secret versions if the values have changed
- Delete old versions

Does it all make sense?

**Review by simoneb — APPROVED** [review]
